### PR TITLE
Parallel Unstructured Mesh Infrastructure (PUMI) integration [pumi-dev]

### PR DESCRIPTION
**Pull request by cwsmith**
This PR adds PUMI support with and without use of the Simmetrix modeling kernel.

#### TODO
- [x] document cmake build process see [README.mfem-pumi-install](https://github.com/mfem/mfem/blob/6ca81275dc86fc8b50cac555e1dbe6b1e0a10cf0/README.mfem-pumi-install)
- [x] support gnu-make build without pumi
- [x] support glvis compilation - need to modify inputs to `MFEMTargets.cmake` and `MFEM_TPL_INCLUDE_DIRS`
- [x] make sure Travis CI and Appveyor tests pass
- [x] update website documentation in https://github.com/mfem/web repo:
    - [x] add a label in the [Examples documentation](http://mfem.org/examples/), see https://github.com/mfem/web/blob/master/src/examples.md
    - [x] mention on the [Features page](http://mfem.org/features/), see https://github.com/mfem/web/blob/master/src/features.md

### CODE REVIEW

File: data/pumi/geom/Kova.dmg [hunk 1]
```diff
@@ -0,0 +1,95 @@
+1 8 18 12
```
Comment (tzanio) [inline data/pumi/geom/Kova.dmg]: We prefer to keep only simple example meshes in `data/` can you move the content of `data/pumi` to the external [mfem/data repo](https://github.com/mfem/data)  instead and reference them as in https://github.com/mfem/mfem/blob/master/miniapps/electromagnetics/joule.cpp#L89-L93?

File: pumi-config.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+cmake \
```
Comment (tzanio) [inline pumi-config.sh]: Remove this file

File: config/cmake/config.hpp.in [hunk 3]
```diff
@@ -98,6 +98,9 @@
 // Enable MFEM functionality based on Conduit
 #cmakedefine MFEM_USE_CONDUIT
 
+// Enable MFEM functionality based on the SCOREC library
```
Comment (tzanio) [inline config/cmake/config.hpp.in]: Is it `SCOREC` library or `SCOREC PUMI` library, or just `PUMI`?

File: config/defaults.mk [hunk 4]
```diff
@@ -106,6 +106,7 @@ MFEM_USE_PETSC       = NO
 MFEM_USE_MPFR        = NO
 MFEM_USE_SIDRE       = NO
 MFEM_USE_CONDUIT     = NO
+MFEM_USE_SCOREC      = NO
```
Comment (tzanio) [inline config/defaults.mk]: Document `MFEM_USE_SCOREC` in the `INSTALL` file.

File: mesh/mesh.hpp [hunk 5]
```diff
@@ -198,6 +198,7 @@ class Mesh
    Element *ReadElement(std::istream &);
    static void PrintElement(const Element *, std::ostream &);
 
+
```
Comment (tzanio) [inline mesh/mesh.hpp]: Revert

File: README.mfem-pumi-install [hunk 6]
```diff
@@ -0,0 +1,286 @@
+The following instructions are for building mfem with pumi on a SCOREC workstation.  Adjust
```
Comment (tzanio) [inline README.mfem-pumi-install]: I suggest we split this in two -- a brief description that goes into `INSTALL` and the more detailed instructions that can go on the webpage, e.g. as a link on http://mfem.org/building

File: CMakeLists.txt [hunk 7]
```diff
@@ -296,6 +303,15 @@ message(STATUS "MFEM build type: CMAKE_BUILD_TYPE = ${CMAKE_BUILD_TYPE}")
 message(STATUS "MFEM version: v${MFEM_VERSION_STRING}")
 message(STATUS "MFEM git string: ${MFEM_GIT_STRING}")
 
+if (SCOREC_FOUND)
+ message(STATUS "MFEM: using package SCOREC")
+ #define a header file with the MFEM_USE_SIMMETRIX preprocessor variable
+ set(MFEM_USE_SIMMETRIX ${SCOREC_gmi_sim_FOUND})
+ configure_file("${CMAKE_CURRENT_SOURCE_DIR}/examples/pumi_config.h.in"
```
Comment (tzanio) [inline CMakeLists.txt]: Can we put `pumi_config.h.in` in the top-level `config` directory instead?

File: examples/pumi_config.h.in [hunk 8]
```diff
@@ -0,0 +1 @@
+#cmakedefine MFEM_USE_SIMMETRIX
```
Comment (tzanio) [inline examples/pumi_config.h.in]: Is this really necessary?

File: examples/GenerateMFEMMesh.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
+//                                PUMI Example
```
Comment (tzanio) [inline examples/GenerateMFEMMesh.cpp]: Should this be a miniapp in `miniapps/meshing` or maybe part of `mesh-explorer` there?

File: examples/pumi_atr_ex2.cpp [hunk 10]
```diff
@@ -0,0 +1,427 @@
+//                                MFEM Example 2 modified
```
Comment (tzanio) [inline examples/pumi_atr_ex2.cpp]: I suggest creating a directory `examples/pumi` and moving `ex1`, `ex1p` and `ex2p` there, following the style of `examples/sundials` and `examples/petsc`.

File: data/pumi/geom/Kova.dmg (hunk 1 above)
Comment (KazemKamran) [reply data/pumi/geom/Kova.dmg]: Done. PUMI examples' `data` is moved to the external [mfem/data/pumi repo](https://github.com/mfem/data/tree/master/pumi)

File: pumi-config.sh (hunk 2 above)
Comment (KazemKamran) [reply pumi-config.sh]: Done.

File: config/cmake/config.hpp.in (hunk 3 above)
Comment (KazemKamran) [reply config/cmake/config.hpp.in]: Corrected. We use `PUMI` library.

File: config/defaults.mk (hunk 4 above)
Comment (KazemKamran) [reply config/defaults.mk]: Documention for `MFEM_USE_PUMI` is added to the `INSTALL` file. 
https://github.com/mfem/mfem/blob/pumi-dev/INSTALL#L352.
We replace all `MFEM_USE_SCOREC` with `MFEM_USE_PUMI`. Following the previous comment.

File: README.mfem-pumi-install (hunk 6 above)
Comment (KazemKamran) [reply README.mfem-pumi-install]: I have added a brief description in the `INSTALL` https://github.com/mfem/mfem/blob/pumi-dev/INSTALL#L352. @cwsmith what do you think about the webpage instructions? I see that this README explains installation on SCOREC machines and LLNL Vulcan. Should we keep it as it is for the website?

File: CMakeLists.txt (hunk 7 above)
Comment (KazemKamran) [reply CMakeLists.txt]: Done.

File: examples/pumi_config.h.in (hunk 8 above)
Comment (KazemKamran) [reply examples/pumi_config.h.in]: I think @cwsmith did it to make addition of SIMMETRIX clean

File: examples/GenerateMFEMMesh.cpp (hunk 9 above)
Comment (KazemKamran) [reply examples/GenerateMFEMMesh.cpp]: I took it out at the moment.

File: examples/pumi_atr_ex2.cpp (hunk 10 above)
Comment (KazemKamran) [reply examples/pumi_atr_ex2.cpp]: Done.

File: mesh/pmesh_pumi.hpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
[the remark this answers is not in the document]
Comment (KazemKamran) [reply mesh/pmesh_pumi.hpp]: We  are putting everything in a one [cpp](https://github.com/mfem/mfem/blob/pumi-dev/mesh/mesh_pumi.cpp) `mesh/mesh_pumi.cpp` and one [header](https://github.com/mfem/mfem/blob/pumi-dev/mesh/mesh_pumi.hpp) file `mesh/mesh_pumi.hpp`.

[the remark this answers is not in the document]
Comment (tzanio) [reply mesh/pmesh_pumi.hpp]: Can you just call it `pumi.?pp`?

File: examples/pumi_config.h.in (hunk 8 above)
Comment (tzanio) [reply examples/pumi_config.h.in]: My question was why is this not in `config/cmake/config.hpp.in` or similar?

File: mesh/pmesh_pumi.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (KazemKamran) [reply mesh/pmesh_pumi.hpp]: No problem. Is a `cpp` and a `hpp` OK or you want just one file?

[the remark this answers is not in the document]
Comment (tzanio) [reply mesh/pmesh_pumi.hpp]: `pumi.hpp` and `pumi.cpp` cf. sundials and petsc

File: miniapps/common/CMakeLists.txt [hunk 12]
```diff
@@ -27,3 +27,6 @@ endif()
 
 add_library(mfem_miniapps_common ${MFEM_MINIAPPS_COMMON_SOURCES}
             ${MFEM_MINIAPPS_COMMON_HEADERS})
+if (MFEM_USE_PUMI)
```
Comment (tzanio) [inline miniapps/common/CMakeLists.txt]: Unless there are changes in `miniapps/`, this shouldn't be necessary.

File: INSTALL [hunk 13]
```diff
@@ -473,6 +480,9 @@ The specific libraries and their options are:
   URL: https://zlib.net
   Options: ZLIB_OPT, ZLIB_LIB.
 
+- PUMI (optional), used when MFEM_USE_PUMI = YES.
+  URL: https://scorec.rpi.edu/pumi
```
Comment (tzanio) [inline INSTALL]: Shouldn't there be `PUMI_OPT` and `PUMI_LIB` variables? How does the regular make work?

File: CHANGELOG [hunk 14]
```diff
@@ -11,8 +11,14 @@
 Version 3.3.3 (development)
 ===========================
 
-More efficient non-conforming adaptive mesh refinement
-------------------------------------------------------
+More general and efficient mesh adaptivity
+------------------------------------------
+- Added support for PUMI, the Parallel Unstructured Mesh Infrastructure from
```
Comment (tzanio) [inline CHANGELOG]: Revise/update the PUMI section if you want.

File: config/pumi_config.h.in [hunk 15]
```diff
@@ -0,0 +1 @@
+#cmakedefine MFEM_USE_SIMMETRIX
```
Comment (tzanio) [inline config/pumi_config.h.in]: Why the separate file? Why can't this be in `config/cmake/config.hpp.in`?

File: fem/lininteg.cpp [hunk 16]
```diff
@@ -735,4 +735,61 @@ void DGElasticityDirichletLFIntegrator::AssembleRHSElementVect(
    }
 }
 
+//L(v) := tau * (f, grad q)
+void StabGradDomainLFIntegrator::AssembleRHSElementVect(const FiniteElement &el,
```
Comment (tzanio) [inline fem/lininteg.cpp]: Do the changes for `StabGradDomainLFIntegrator` need to be part of this PR? (I'm not opposed to adding new integrators, but maybe in a separate PR).

File: config/pumi_config.h.in (hunk 15 above)
Comment (cwsmith) [reply config/pumi_config.h.in]: Sure, I'll see if we can use `config.hpp.in`.

File: INSTALL (hunk 13 above)
Comment (cwsmith) [reply INSTALL]: This merge doesn't support gnu-make builds; just cmake.  We checked to see that we did not break the mfem gnu-make build (without pumi).

File: miniapps/common/CMakeLists.txt (hunk 12 above)
Comment (cwsmith) [reply miniapps/common/CMakeLists.txt]: OK.  We'll remove this.

File: fem/lininteg.cpp (hunk 16 above)
Comment (KazemKamran) [reply fem/lininteg.cpp]: No problem. We can add it in pumi-NS merge.

File: CHANGELOG (hunk 14 above)
Comment (KazemKamran) [reply CHANGELOG]: This looks great. Thank you.

File: miniapps/common/CMakeLists.txt (hunk 12 above)
Comment (cwsmith) [reply miniapps/common/CMakeLists.txt]: Removing

```
if (MFEM_USE_PUMI)
 target_link_libraries(mfem_miniapps_common PUBLIC SCOREC::core)
endif()
```

 results in a compilation error:

```
[ 32%] Building CXX object CMakeFiles/mfem.dir/fem/hybridization.cpp.o
In file included from /lore/cwsmith/develop/mfem/mfem/c:40:0,
                from /lore/cwsmith/develop/mfem/mfem/mfem.hpp:38,
                from /lore/cwsmith/develop/mfem/mfem-build/mfem.hpp:3,
                from /lore/cwsmith/develop/mfem/mfem/miniapps/common/fem_extras.hpp:15,
                from /lore/cwsmith/develop/mfem/mfem/miniapps/common/fem_extras.cpp:12:
/lore/cwsmith/develop/mfem/mfem/mesh/pumi.hpp:39:18: fatal error: pumi.h: No such file or directory
#include <pumi.h>
                 ^
compilation terminated.
```

I don't think we can remove this `target_link_libraries` (which brings in the compiler flag to include the pumi install directory) call without changes to where the pumi headers are included.

File: config/pumi_config.h.in (hunk 15 above)
Comment (cwsmith) [reply config/pumi_config.h.in]: fixed https://github.com/mfem/mfem/pull/237/commits/6601d3757b278a069dc6afb9bc72463f52d986ed

File: miniapps/common/CMakeLists.txt (hunk 12 above)
Comment (v-dobrev) [reply miniapps/common/CMakeLists.txt]: @cwsmith, it looks like when `MFEM_USE_PUMI` is enabled, `SCOREC::core` needs to be a dependency of the library, `mfem`. That way the dependency is propagated to everything that depends on `mfem`.

Comment (cwsmith) [reply miniapps/common/CMakeLists.txt]: @v-dobrev Thank you.  SCOREC::core was already set as a dependency of mfem:
https://github.com/mfem/mfem/blob/1bd99fd5dc0541e60c081f4330e61788d9b0bdb5/CMakeLists.txt#L312

but, digging a bit more, and it looks like `mfem_miniapps_common` does not depend on `mfem`.  Removing:
https://github.com/mfem/mfem/blob/1bd99fd5dc0541e60c081f4330e61788d9b0bdb5/miniapps/common/CMakeLists.txt#L30-L32
and adding the following line
```
target_link_libraries(mfem_miniapps_common PUBLIC mfem)
```
gets me through the compile.  I'm not sure if this is a cleaner solution though.

Comment (v-dobrev) [reply miniapps/common/CMakeLists.txt]: OK, pick the one that you prefer.

Comment (cwsmith) [reply miniapps/common/CMakeLists.txt]: @v-dobrev OK.  Let's stick with what is there now (`target_link_libraries(... SCOREC::core)`).  If the other packages hit a similar issue we could reconsider using `target_link_libraries(... mfem)`.

File: INSTALL [hunk 17]
```diff
@@ -473,7 +473,6 @@ The specific libraries and their options are:
   URL: https://zlib.net
   Options: ZLIB_OPT, ZLIB_LIB.
 
-
```
Comment (tzanio) [inline INSTALL]: Can we bring back the changes in this file... we will eventually like to make the pure Makefile system work too :)

Comment (KazemKamran) [reply INSTALL]: @tzanio as @cwsmith  commented before, this merge doesn't support gnu-make builds with pumi enabled; just cmake. We did check to see that we did not break the mfem gnu-make build (without pumi). That was my fault to modifying INSTALL in the first place.

File: CHANGELOG [hunk 18]
```diff
@@ -11,8 +11,14 @@
...
+  https://scorec.rpi.edu/pumi. PUMI is an unstructured, distributed mesh data
+  management system that is capable of handling general non-manifold models and
+  effectively supports automated adaptive analysis. PUMI enables for the first
+  time support for parallel unstructured modifications of MFEM meshes.
+
 - Significantly reduced MPI communication in the construction of the parallel
```
Comment (tzanio) [inline CHANGELOG]: @KazemKamran: can you add a blurb describing the new examples in the "New and updated examples and miniapps" section, see for example the one for the NURBS miniapps.

File: examples/pumi/ex1.cpp [hunk 19]
```diff
@@ -0,0 +1,247 @@
+//                                MFEM Example 1
```
Comment (tzanio) [inline examples/pumi/ex1.cpp]: @KazemKamran: can you unify the example headers to be of this form, see e.g. the files inexamples/sundials and examples/petsc?

File: examples/pumi/ex2p.cpp [hunk 20]
```diff
@@ -0,0 +1,429 @@
+//                                MFEM Example 2 modified
```
Comment (tzanio) [inline examples/pumi/ex2p.cpp]: Should this be ex2.cpp (serial) or ex2p.cpp (parallel)?

File: examples/pumi/ex6p.cpp [hunk 21]
```diff
@@ -0,0 +1,381 @@
+//                       PUMI Example 6 - Parallel Version
```
Comment (tzanio) [inline examples/pumi/ex6p.cpp]: TODO: add PUMI examples section in `doc/CodeDocumentation.dox` (similar to SUNDIALS and PETSc)

File: mesh/pumi.cpp [hunk 22]
```diff
@@ -0,0 +1,1909 @@
...
+   unsigned int j=0;
+   while ((ent = apf_mesh->iterate(itr)))
+   {
+      // Get vertices
+      apf::Downward verts;
+      // int num_vert =
+      apf_mesh->getDownward(ent,0,verts);
+      // Get attribute Tag vs Geometry
+      int attr = 1;
+      /*if (apf_mesh->hasTag(ent,atts)){
```
Comment (tzanio) [inline mesh/pumi.cpp]: @KazemKamran: can you please clean up commented/unused code like this before we merge?

File: examples/pumi/ex1.cpp [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+#include <crv.h>
+
+using namespace std;
+using namespace mfem;
+
+int main(int argc, char *argv[])
+{
+   // 1. Initilize MPI
+   int num_proc, myId;
+   MPI_Init(&argc, &argv);
```
[the remark this answers is not in the document]
Comment (KazemKamran) [reply examples/pumi/ex1.cpp]: "Yes - pumi always requires mpi" provided by @cwsmith  :)

[the remark this answers is not in the document]
Comment (tzanio) [reply examples/pumi/ex1.cpp]: Ok, so then what's the value of having ex1.cpp and ex1p.cpp? Isn't the former just a version of the latter on 1 MPI task?

(In MFEM, the value is that ex1.cpp can be build and run without any external dependency: MPI, hypre, etc.)

File: examples/pumi/makefile [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+%-test-par: %
+	@$(call mfem-test,$<, $(RUN_MPI), $(PARALLEL_NAME))
+%-test-seq: %
+	@$(call mfem-test,$<,, $(SERIAL_NAME))
+
+# Testing: Example-specific execution options:
+ex1-test-par: ex1
+	@$(call mfem-test,$<, $(RUN_MPI_NP) 1, $(PARALLEL_NAME))
+ex1p-test-par: ex1p
+	@$(call mfem-test,$<, $(RUN_MPI_NP) 8, $(PARALLEL_NAME))
```
[the remark this answers is not in the document]
Comment (KazemKamran) [reply examples/pumi/makefile]: Well the mesh that we have provided for this example is partitioned in 8. (https://github.com/mfem/data/tree/master/pumi/parallel/Kova)

File: examples/pumi/makefile [hunk 25]
```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at the
```
[the remark this answers is not in the document]
Comment (KazemKamran) [reply examples/pumi/makefile]: OK, Sorry.

**Comment by v-dobrev** [comment]
I think this PR is ready to be merged. Thanks, @KazemKamran and @cwsmith!

Can you guys confirm that everything still works OK on your side, after the changes I made?

File: CHANGELOG (hunk 18 above)
Comment (KazemKamran) [reply CHANGELOG]: Done

File: examples/pumi/ex1.cpp (hunk 19 above)
Comment (KazemKamran) [reply examples/pumi/ex1.cpp]: Done

**Comment by tzanio** [comment]
@KazemKamran, when you are done and happy can you approve the PR?

**Comment by tzanio** [comment]
Nice work @KazemKamran and @cwsmith, thanks!

**Comment by tzanio** [comment]
Also update website documentation: https://github.com/mfem/web/pull/12

**Pull request merged**